### ISSUE TITLE
Extremely bad tick performance - food and water seeking?

### ISSUE THREAD

**Issue by Protected** [capnkirok/animania#490]
Minecraft 1.12.2
Forge 14.23.1.2555
Animania 1.4.3

With "default" settings (animals don't starve to death but require food and water to breed), in my server, some time after any player's farm with a few dozen animals is loaded (the chunks are loaded due to someone being around) the server starts skipping hundreds of ticks at a time. After profiling, animania animals seemed to be taking several seconds per tick. I ran some tests with no one else around and it appeared to me that everything was originally fine (when all the animals were well fed), but if the animals got hungry/thirsty and food or water were not immediately available, server ticking slowed to a crawl (possibly due to animals attempting to move towards nearby farms but being prevented by fences?) To confirm this issue, I changed to "ambient mode" (animals don't require food and water any more) and was no longer able to reproduce the problem with a test set of over 100 breeding cows and bulls in the same chunk. What do you think?

**Comment by capnkirok** [comment]
Hey @Protected yes, sadly, I think you are correct. It doesn't happen for everyone, and only seems to occur on servers with many other mods. But it ruins the experience.

I can't tell you how much time I've put into this issue, but I believe I need to rewrite the whole thing again from scratch. It becomes quite noticeable with so many animals in the chunk. Going to leave this one open. Recommend you keep it in Ambiance mode for now.

**Comment by Protected** [comment]
Hi, thanks for the quick reply. It's in ambient mode but players were complaining because of: a) Explosive amount of breeding (there's already an open issue for that) can't be controlled if the animals can just breed freely and b) apparently milk still requires water but cows will no longer drink of their own volition in this mode (I can open a separate issue for that if you think it's necessary, but I suppose it's just how it is with these settings) so I was hoping I'd be able to go back to eating and drinking mode. But I understand this can be difficult to pinpoint, so we'll patiently stick with ambient mode for now!

We have fewer mods than most. We made our own modpack and since half of our (private server's players) dislike mods entirely, we're using mostly mods that don't disrupt vanilla gameplay; animania is one of the riskiest! Here's a list of mods on the server:

```
animania-1.12.2-1.4.3.jar
AppleSkin-mc1.12-1.0.9.jar
Authenticate-1.0.jar
badwithernocookiereloaded-1.12-2.12.jar
Baubles-1.12-1.5.1.jar
BiblioCraft[v2.4.3][MC1.12.0].jar
calculator-1.12.2-5.0.1.jar
ChestTransporter-1.12-2.8.7.jar
chiselsandbits-14.11.jar
Chunk Pregenerator V1.12-1.5.1.jar
CodeChickenLib-1.12-3.1.3.313-universal.jar
CoFHCore-1.12.2-4.3.7.25-universal.jar
CoFHWorld-1.12.2-1.0.2.10-universal.jar
CookingForBlockheads_1.12.1-6.3.17.jar
CraftStudio-1.0.0.93-mc1.12-alpha.jar
creeperconfetti-1.4.2.jar
CTM-MC1.12-0.2.3.9.jar
DiscordChat-1.12.2-2.3.0.jar
DoggyTalents-1.12-1.14.2.314-universal.jar
DynamicSurroundings-1.12.2-3.4.7.2.jar
fluxnetworks-1.12.2-3.0.1.jar
Hats-1.12.2-7.0.2.jar
iChunUtil-1.12.2-7.1.3.jar
ImmersiveEngineering-0.12-75-core.jar
ImmersiveEngineering-0.12-75.jar
InventoryTweaks-1.63.jar
IvToolkit-1.3.3-1.12.jar
jei_1.12.2-4.8.5.132.jar
journeymap-1.12.2-5.5.2.jar
JustEnoughResources-1.12.2-0.8.5.27.jar
mcjtylib-1.12-2.5.0.jar
MCMultiPart-2.3.3.jar
MobRules-1.12.2-1.0.0.jar
modnametooltip_1.12.2-1.10.0.jar
NaturesCompass-1.12.2-1.5.0.jar
NetherPortalFix_1.12.1-5.3.13.jar
OpenGlider-1.12.1-1.1.0.jar
RecurrentComplex-1.4.6.jar
RedstoneFlux-1.12-2.0.1.2-universal.jar
Reliquary-1.12.2-1.3.4.724.jar
rftools-1.12-7.14.jar
sonarcore-1.12.2-5.0.2.jar
supersoundmuffler-1.12.1-1.0.2.9.jar
theoneprobe-1.12-1.4.19.jar
ThermalDynamics-1.12.2-2.3.7.14-universal.jar
ThermalExpansion-1.12.2-5.3.7.31-universal.jar
ThermalFoundation-1.12.2-2.3.7.18-universal.jar
TickProfiler-1.12-0.0.1.jar
Waystones_1.12.1-4.0.17.jar
WearableBackpacks-1.12.2-3.1.0.jar
xnet-1.12-1.4.1.jar
```

Some notes on mods you probably don't have to worry about:

- Authenticate is my own quickly cobbled together mod for bridging minecraft user accounts with users on our Discord. It does not do anything to entities.
- Chunk pregenerator doesn't do anything unless I tell it to.
- Discord chat is likewise probably harmless in this case.
- JEI/JER: Probably safe?
- Journeymap: Probably safe?
- MobRules is my rules-based filter for entity spawning. It is not configured to interfere with animals in any way.
- modnametooltip: Probably safe?
- TOP: Probably safe?
- TickProfiler: Definitely safe (only added to help diagnose this issue)

Some mods that might interfere with animania:

- DoggyTalents and Hats definitely do things to animals.
- DynamicSurroundings might do things to animals.
- Some of the advanced RF mods (I'm not 100% familiar with them).

Are there any common mods across everyone with this issue?

**Comment by Baughn** [comment]
Dynamic Surroundings should be client-side only. At least, I'm running a server without it installed, and it still appears to work fine on the client -- given its nature, I'd suspect it simply disables itself on the server.

I've got the same problem, of course, though using a modpack that's an expansion of FTB Revelation. Tons of mods. What I can report is that turning on ambient mode negates the problem, at the cost of animals breeding all on their own.

What's the difference between ambient mode on and off? If it's just pathing to food, they shouldn't need to do that every tick... even if they're hungry/thirsty, there's no point in looking for food *every* tick.

Assuming it actually does that, then scheduling attempted feedings for once every minute or so should effectively remove the problem.

**Comment by Baughn** [comment]
As for the list of mods above, the ones it shares with my server are:

- Animania (of course).
- Appleskin
- badwithernocookiereloaded-1.12-2.12.jar
- Baubles-1.12-1.5.1.jar
- BiblioCraft[v2.4.3][MC1.12.0].jar
- ChestTransporter-1.12-2.8.7.jar
- chiselsandbits-14.11.jar
- CodeChickenLib-1.12-3.1.3.331-universal.jar
- CoFHCore-1.12.2-4.3.10.5-universal.jar
- CoFHWorld-1.12.2-1.1.1.12-universal.jar
- CookingForBlockheads_1.12.1-6.3.17.jar
- CraftStudio-1.0.0.93-mc1.12-alpha.jar
- CTM-MC1.12-0.2.3.9.jar
- fluxnetworks-1.12.2-3.0.2.jar
- iChunUtil-1.12.2-7.1.4.jar
- ImmersiveEngineering-0.12-76-core.jar
- ImmersiveEngineering-0.12-75.jar
- InventoryTweaks-1.64-dev.jar
- IvToolkit-1.3.3-1.12.jar
- jei_1.12.2-4.8.5.132.jar
- journeymap-1.12.2-5.5.2.jar
- JustEnoughResources-1.12.2-0.8.6.33.jar
- mcjtylib-1.12-2.6.2.jar
- MCMultiPart-2.4.1.jar
- modnametooltip_1.12.2-1.10.0.jar
- NaturesCompass-1.12.2-1.5.1.jar
- NetherPortalFix_1.12.1-5.3.13.jar
- RecurrentComplex-1.4.7.jar
- RedstoneFlux-1.12-2.0.1.2-universal.jar
- Reliquary-1.12.2-1.3.4.728.jar
- rftools-1.12-7.23.jar
- sonarcore-1.12.2-5.0.4.jar
- theoneprobe-1.12-1.4.19.jar
- ThermalDynamics-1.12.2-2.3.10.4-universal.jar
- ThermalExpansion-1.12.2-5.3.10.14-universal.jar
- ThermalFoundation-1.12.2-2.3.10.6-universal.jar
- TickProfiler-1.12-0.0.2.jar
- Waystones_1.12.1-4.0.17.jar
- WearableBackpacks-1.12.2-4.0.20.jar
- xnet-1.12-1.6.7.jar


This leaves the following mods as probably not the cause:

- calculator-1.12.2-5.0.1.jar
- Chunk Pregenerator V1.12-1.5.1.jar
- creeperconfetti-1.4.2.jar
- DiscordChat-1.12.2-2.3.0.jar
- DoggyTalents-1.12-1.14.2.314-universal.jar
- DynamicSurroundings-1.12.2-3.4.7.2.jar
- Hats-1.12.2-7.0.2.jar
- MobRules-1.12.2-1.0.0.jar
- OpenGlider-1.12.1-1.1.0.jar
- supersoundmuffler-1.12.1-1.0.2.9.jar

...but the only one which plausibly could have been was Doggy Talents.

**Comment by capnkirok** [comment]
Yeah, the pathing does check at intervals. However, the current issue is a result of:
- Animal identifying a block/tile entity source that it can't get to
- Animal caught between two equidistant sources

In either of these cases, the animal can overindex and repeat the pathing set, which is time intensive.

You can also enable 'manual' breeding if you are seeing too much breeding.

In any case, the fix appears this weekend, so you won't have to wait too long.

**Comment by capnkirok** [comment]
Great news! After 9 months of this issue plaguing me, I found the problem and it will be fixed in the next release.

For those wondering what the issue was.... the overindexing was a result of where I had placed the execution code in the AI, which caused it to repeatedly reset pathing if it didn't fulfill its goal.

This makes a huge difference in performance. Update should be out today or tomorrow. Thank you for your patience!
